Update to 4.1 anonymous function syntax (#686)
AlexsLemonade/training-modules: scRNA-seq/05-clustering_markers_scRNA.Rmd

This returns a list of data frames, which isn't quite what we want.

There is no built-in function that will give us just the first few _row names_, so we will have to define one.

> As of version 4.1, R introduced a new approach to defining _anonymous functions_ - that is, functions you can quickly define "on-the-fly" without formally assigning them.

Reviewer: Ending the sentence with "formally assigning them" seemed odd. I don't like my solution either, tbh, but it is maybe something. 
```suggestion
As of version 4.1, R introduced a new approach to defining _anonymous functions_ - that is, functions you can quickly define "on-the-fly" without formally assigning them to a function name.
```

They are handy when you need to do a very short task that requires a function, but it isn't really a function you need beyond this context.
This new anonymous syntax looks like this: `\(x)...` (or for slightly longer code, use curly braces as in `\(x) {...}`).

> This would define an anonymous function that takes one argument, `x`, and `...` is where you would put the expression to calculate, aka the "body" of the function.

Reviewer: I started by removing the repetition of `anonymous` here, but I moved into a bit more revision. I didn't think the "a.k.a." really added much, so I just removed that. (I also wasn't sure what "aka" was referring to, as I don't think there was really another name there!)
```suggestion
This defines a function that takes one argument, `x`, with `...` indicating where you would put the expression to calculate.
```

`purrr::map()` will then apply the expression in our anonymous function to each element of the list, and return the results as a new list.

```{r head_markernames, live = TRUE}
# Get the first few row names of each table with a purrr function.
purrr::map(as.list(markers), # convert markers to a 'regular' list for purrr

>            \(x) rownames(head(x))) # our custom function.

Reviewer: Maybe... I do think we should reverse the order of the two functions applied, but feel less certain about whether adding a pipe in the middle helps. 
```suggestion
           \(x) rownames(x) |> head() ) # our custom function.
```

Author: I think for this I prefer no pipe since the syntax is already busy with symbols. But definitely `head( rownames(x) )` in that order!

```

Another variant is `purrr::imap()`, which allows us to use the names of the list elements in our function.
(Try `names(markers)` to see the names for the list we are working with now.)
We will use that here to name output files where we will print each of the marker tables, one for each cell type.
We are again defining a custom function within the call to `purrr:imap()` using the `\(x)` syntax, but this time we can use `x` for the list elements and `y` for their names.

> So, we'll actually start by defining the function as `\(x,y)`, since there will be two input arguments.

Author: ```suggestion
So, we'll actually start by defining the function as `\(table, id)`, since there will be two input arguments.
```
